sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -149,69 +162,122 @@ Flux<ByteBuffer> decryptBlob(String encryptedDataString, Flux<ByteBuffer> encryp
+                            }
 
-                    Cipher cipher;
-                    try {
-                        cipher = getCipher(contentEncryptionKey, encryptionData, iv, padding);
-                    } catch (InvalidKeyException e) {
-                        throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                            return encryptedFlux.map(encryptedByteBuffer -> {
+                                /*
+                                 * If we could potentially decrypt more bytes than encryptedByteBuffer can hold, allocate more
+                                 * room. Note that, because getOutputSize returns the size needed to store
+                                 * max(updateOutputSize, finalizeOutputSize), this is likely to produce a ByteBuffer slightly
+                                 * larger than what the real outputSize is. This is accounted for below.
+                                 */
+                                ByteBuffer plaintextByteBuffer = ByteBuffer
+                                    .allocate(cipher.getOutputSize(encryptedByteBuffer.remaining()));
+
+                                // First, determine if we should update or finalize and fill the output buffer.
+                                int bytesToInput = encryptedByteBuffer.remaining();
+                                try {
+                                    // We will have reached the end of the downloaded range, finalize.
+                                    if (totalInputBytes.longValue() + bytesToInput
+                                        >= encryptedBlobRange.getAdjustedDownloadCount()) {
+                                        cipher.doFinal(encryptedByteBuffer, plaintextByteBuffer);
+                                    } else {
+                                        // We won't reach the end of the downloaded range, update.
+                                        cipher.update(encryptedByteBuffer, plaintextByteBuffer);
+                                    }
+                                } catch (GeneralSecurityException e) {
+                                    throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+                                }
+                                totalInputBytes.addAndGet(bytesToInput);
+
+                                // Flip the buffer to set the position back to 0 and set the limit to the data size.
+                                plaintextByteBuffer.flip();
+                                return plaintextByteBuffer;
+                            });
+                        case ENCRYPTION_PROTOCOL_V2:
+                            // Buffer an exact region with the nonce and tag
+                            BufferStagingArea stagingArea =
+                                new BufferStagingArea(GCM_ENCRYPTION_REGION_LENGTH + TAG_LENGTH + NONCE_LENGTH,
+                                GCM_ENCRYPTION_REGION_LENGTH + TAG_LENGTH + NONCE_LENGTH);
```

Comment:
We should probably be reading these values from the JSON

Reply:
Do we need to? Didn't we say the values in the json string are so someone else not using our clients could decrypt them? They are constants for v2, so it shouldn't matter?